### PR TITLE
Fix suspense fallback is invalid

### PR DESCRIPTION
**Pull request by lamhieu-vk**
Resolved https://github.com/airbnb/enzyme/issues/2200

I updated `enzyme-adapter-react-16` to fix issue suspense fallback is invalid. I handled to always wrap `React.Suspense` component outside of `React.Lazy` and `React.Fallback` components in any case.

I passed all tests in master branch and local test by below code:
```js
test('renders suspenseFallback=true', () => {
  const el = <ComponentWrapper />;
  const wrapper = shallow(el, { suspenseFallback: true });
  expect(wrapper.find(Fallback).length).toEqual(1);
  expect(wrapper.find(LazyComponent).length).toEqual(0);
});

test('renders suspenseFallback=false', () => {
  const el = <ComponentWrapper />;
  const wrapper = shallow(el, { suspenseFallback: false });
  expect(wrapper.find(Fallback).length).toEqual(0);
  expect(wrapper.find(LazyComponent).length).toEqual(1);
});
```
Code cloned from issue to reproduce, it's failed in master branch and passed in this branch.

Note: Lazy component will always display as `<lazy />` in debug and If you want to expect this component, should be use `.find(Component)` instead of `.find('ComponentName')`.

Thanks for review!

### CODE REVIEW

File: packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js [hunk 1]
```diff
@@ -120,6 +120,26 @@ function unmemoType(type) {
...
 
+function transformSuspenseElement(el, { suspenseFallback }) {
+  let { children } = el.props;
+
+  if (suspenseFallback) {
+    const { fallback } = el.props;
+    children = replaceLazyWithFallback(children, fallback);
+  }
+
+  const FakeSuspenseWrapper = (props) => React.createElement(el.type, { ...el.props, ...props }, children);
```
Comment (ljharb) [inline packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js]: this seems like it might break class components that use private fields, or functional components that use hooks.

File: packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js [hunk 2]
```diff
@@ -256,8 +276,9 @@ function toTree(vnode) {
         rendered: childrenToTree(node.child),
       };
     }
-    case FiberTags.Lazy:
+    case FiberTags.Lazy: {
```
Comment (ljharb) [inline packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js]: the curlies are only needed when there's block-scoped variables inside the case; let's revert this.

File: packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js [hunk 3]
```diff
@@ -528,7 +549,7 @@ class ReactSixteenAdapter extends EnzymeAdapter {
       }
       if (lastComponent !== Component) {
         if (isStateful(Component)) {
-          wrappedComponent = class extends Component {}; // eslint-disable-line react/prefer-stateless-function
+          wrappedComponent = class extends Component { }; // eslint-disable-line react/prefer-stateless-function
```
Comment (ljharb) [inline packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js]: please revert unrelated formatting changes
```suggestion
          wrappedComponent = class extends Component {}; // eslint-disable-line react/prefer-stateless-function
```

File: packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js [hunk 4]
```diff
@@ -588,6 +609,22 @@ class ReactSixteenAdapter extends EnzymeAdapter {
       return wrappedComponent;
     };
 
+    const renderElement = (elConfig, ...rest) => {
+      const renderedEl = renderer.render(elConfig, ...rest);
+
+      const typeIsExisted = !!(renderedEl && renderedEl.type);
```
Comment (ljharb) [inline packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js]: when would this not be true?

File: packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx [hunk 5]
```diff
@@ -1722,7 +1722,7 @@ describe('shallow', () => {
     });
 
     it('finds LazyComponent when render component wrapping lazy component', () => {
-      const LazyComponent = lazy(fakeDynamicImport(DynamicComponent));
+      const LazyComponent = lazy(() => fakeDynamicImport(DynamicComponent));
```
Comment (ljharb) [inline packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]: why are these test changes needed?

File: packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js (hunk 2 above)
Comment (lamhieu-vk) [reply packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js]: I reverted it!

File: packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js (hunk 3 above)
Comment (lamhieu-vk) [reply packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js]: I reverted it!

File: packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js (hunk 4 above)
Comment (lamhieu-vk) [reply packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js]: Yep, sometimes `type` in Element is `undefined ` with `$$typeof` is `(react.element)`. I'm not sure I know which cases it can be `undefined`.

File: packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx (hunk 5 above)
Comment (lamhieu-vk) [reply packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]: I changes to like user cases example in [react docs](https://reactjs.org/docs/code-splitting.html#reactlazy).

```js
const OtherComponent = React.lazy(() => import('./OtherComponent'));
```

> React.lazy takes a function that must call a dynamic import(). This must return a Promise which resolves to a module with a default export containing a React component.

File: packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js (hunk 1 above)
Comment (lamhieu-vk) [reply packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js]: In all cases `el.type` is `react.Suspense`, the wrapper of `lazy` and `fallback` inside. I just create it after changed children inside.

File: packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js (hunk 4 above)
Comment (ljharb) [reply packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js]: We've got a bunch of existing helpers for this kind of thing; is there one of those we could use here?

File: packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js (hunk 1 above)
Comment (ljharb) [reply packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js]: Perhaps it would be cleaner to inline `transformSuspenseElement`, since it's only used once, and then it couldn't possibly be invoked when `el` isn't a Suspense element?

File: packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx (hunk 5 above)
Comment (ljharb) [reply packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]: I'm concerned about changing existing tests; instead, let's leave these unchanged, and also *add* test cases with the thunk?

Comment (lamhieu-vk) [reply packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]: I viewed `lazy` test of react [here](https://github.com/facebook/react/blob/b979db4e7215957f03c4221622f0b115a868439a/packages/react-reconciler/src/__tests__/ReactLazy-test.internal.js#L37). The old test is not correct with user case of react `lazy` but I don't know it still passed. However, I think we should update it!

File: packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js (hunk 4 above)
Comment (lamhieu-vk) [reply packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js]: I don't see any helper to check is existed type in element. Do you want to me create a helper to use it?

File: packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js (hunk 1 above)
Comment (lamhieu-vk) [reply packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js]: I updated it, make inline `transformSuspenseElement` function. Thanks for feedback!

**Comment by lamhieu-vk** [comment]
hi @ljharb , could you please give me a feedback?

**Comment by lamhieu-vk** [comment]
hi @ljharb, PR is ready merge!

**Comment by cascornelissen** [comment]
Looks like this change broke the `.instance()` method on a `shallow()` rendered component for us when the root component is a `React.Suspense`.

```
const wrapper = shallow(<ComponentWithReactSuspense />);
wrapper.instance(); // Always null
```

Any ideas?
